scipy/cluster/hierarchy.py

```diff
@@ -995,7 +1002,8 @@ def linkage(y, method='single', metric='euclidean', optimal_ordering=False):
     >>> dn = dendrogram(Z)
     >>> plt.show()
     """
-    y = np.asarray(y, order='C', dtype=np.float64)
+    xp = array_namespace(y)
+    y = as_xparray(y, order='C', dtype=xp.float64, xp=xp)
```

Comment:
I'm working on gh-21835 and this confused me, since `order` only works for NumPy arrays.

It looks like `order='C'` was added in https://github.com/scipy/scipy/commit/57e1ec714bbcbb2ce6a4bbc76362b5071e294890 to ensure that the data is contiguous before it goes into compiled code. But it seems like non-contiguous arrays from other libraries will get past this and not be forced to be contiguous when they are turned into NumPy arrays. 

It this still important? Or at this point, does our machinery recognize that the arrays are not contiguous and copy the data to be contiguous (and in the appropriate order) before passing it into compiled code?

Should there be an error when the array library does not support options passed into the `_asarray` function?